_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,610 @@
+          assert Succ(elem.e) in someset;
+        }
+      } else if elem in preds {
+        assert elem in someset by {
+          assert elem.Pred?;
+          assert elem.e in S(i-1);
+          assert elem.e in someset;
+          assert Pred(elem.e) in someset;
+        }
+      } else if elem in iszeros {
+        assert elem in someset by {
+          assert elem.IsZero?;
+          assert elem.e in S(i-1);
+          assert elem.e in someset;
+          assert IsZero(elem.e) in someset;
+        }
+      } else {
+        assert elem in ifs;
+        assert elem in someset by {
+          assert elem.cond in S(i-1) && elem.thn in S(i-1) && elem.els in S(i-1);
+          assert elem.cond in someset && elem.thn in someset && elem.els in someset;
+          assert If(elem.cond, elem.thn, elem.els) in someset;
+        }
+      }
+    }
+  }
+}
+{% endhighlight %}
+
+## 3. The constructive set is included in the smallest inductive set that satisfies the induction criteria
+
+We can deduce from the previous result that the constructive definition of all terms is also included in any set of term that satisfies the induction criteria. From this we can deduce automatically that the constructive definition of all terms is included in the smallest inductive set satisfying the induction criteria.
+
+{% highlight javascript %}
+// 3.2.6.b.1 AllTermsConstructively is a subset of any set satisfying the induction criteria (hence AllTermsConstructively <= AllTermsInductively)
+lemma AllTermsConstructivelyIncludedInSetsSatisfyingInductionCriteria(
+  terms: iset<Term>
+)
+  requires InductionCriteria(terms)
+  ensures AllTermsConstructively <= terms
+{
+  forall i: nat ensures S(i) <= terms {
+    InductionCriteriaHasConcreteIAsSubset(i, terms);
+  }
+}
+
+lemma AllTermsConstructivelyIncludedInAllTermsInductively()
+  ensures AllTermsConstructively <= AllTermsInductively
+{
+  InductiveAxioms();
+  AllTermsConstructivelyIncludedInSetsSatisfyingInductionCriteria(AllTermsInductively);
+}
+{% endhighlight %}
+
+## 4. Conclusion with the equality
+
+Because we have `<=` and `>=` between these two sets, we can now prove equality.
+
+{% highlight javascript %}
+lemma InductionAndConcreteAreTheSame()
+  ensures AllTermsConstructively == AllTermsInductively
+{
+  InductiveAxioms();
+  AllTermsConstructivelySatisfiesInductionCriteria();
+  AllTermsConstructivelyIncludedInAllTermsInductively();
+}
+{% endhighlight %}
+
+## Bonus Conclusion
+
+We were able to put together two definitions for infinite sets, and prove that these sets were equivalents.
+As stated in the introduction, having multiple definitions of a single infinite set makes it possible to pick the definition adequate to the job to prove other results. For example,
+
+- If a term is in the constructive set, then it cannot be `Add` for example, because it would need to be in a `S(i)` and none of the `S(i)` define `Add`. This can be illustrated in Dafny with the following lemma:
```

Comment:
```suggestion
- If a term is in the constructive set, then it cannot use/(or something similar) `Add` for example, because it would need to be in a `S(i)` and none of the `S(i)` define `Add`. This can be illustrated in Dafny with the following lemma:
```

Reply:
```suggestion
- If a term is in the constructive set, then it cannot be constructed with `Add` for example, because it would need to be in a `S(i)` and none of the `S(i)` define `Add`. This can be illustrated in Dafny with the following lemma:
```